src/Store.ts

```diff
@@ -0,0 +1,392 @@
+	 * multi-key identities and object's whose properties are not stored directly on the object.
+	 *
+	 * @param object The target object
+	 * @param identityArg The argument used to set the identity
+	 */
+	protected _setIdentity(object: { [ index: string ]: any; set: (...args: any[]) => any}, identityArg: string | number) {
+		if (object.set) {
+			object.set(this.idProperty, identityArg);
+		} else {
+			object[this.idProperty] = identityArg;
+		}
+	}
+
+	/**
+	 * This creates a new instance from the store's model.
+	 * @param properties The properties that are passed to the model constructor to
+	 * be copied onto the new instance. Note, that should only be called
+	 * when new objects are being created, not when existing objects
+	 * are being restored from storage.
+	 */
+	create(properties: {}) {
+		return new this.Model(properties);
+	}
+
+	/**
+	 * Retrieves a range of objects from the collection, returning a promise to an array.
+	 *
+	 * @param args Contains the starting index of objects to return (0-indexed)
+	 * and the exclusive end of objects to return
+	 * @return A FetchPromise that will resolve with the results of the fetch
+	 */
+	fetchRange(args: dstore.FetchRangeArgs):dstore.FetchPromise<T> {
+		throw new Error("This method is abstract");
+	}
+
+	filter<T>(...args: any[]): dstore.Collection<T> {
+		return QueryMethod(<QueryMethodArgs<T>> {
+			type: 'filter',
+			normalizeArguments: function (filter) {
+				const Filter = this.Filter;
+				if (filter instanceof Filter) {
+					return [ filter ];
+				}
+				return [ new Filter(filter) ];
+			}
+		}).apply(this, args);
+	}
+
+	/**
+	 * Iterates over the query results, based on
+	 * https://developer.mozilla.org/en/Core_JavaScript_1.5_Reference/Objects/Array/forEach.
+	 * Note that this executes asynchronously, and returns a promise to indicate when the
+	 * callback has been executed.
+	 *
+	 * @param callback Function that is called for each object in the query results
+	 * @parma thisObject The object to use as |this| in the callback.
+	 * @returns A Promise indicating when the operatio is done
+	 */
+	forEach<T>(callback: (item: T, index: number | string, collection: T[]) => void, thisObject?: any) {
+		return this.fetch().then((data) => {
+			if (data) {
+				let i: number;
+				let item: any;
+				for (i = 0; (item = data[i]) !== undefined; i++) {
+					callback.call(thisObject, item, i, this);
+				}
+			}
+			return data;
+		});
+	}
+
+	/**
+	 * Returns an object's identity
+	 * @param object The object to get the identity from
+	 * @returns A string or number representing the identity of the object
+	 */
+	getIdentity(object: { [ name: string ]: any, get?: (name: string) => any }) {
+		return object.get ? object.get(this.idProperty) : object[this.idProperty];
+	}
+
+	emit(event: EventObject) {
+		return this.storage.emit(event);
+	}
+
+	on(type: string, listener: (event: { type: string, beforeId: string | number }) => void) {
```

Comment:
Yes I think so, in fact with the update to TS 1.6  not using ChangeEvent with an optional property was causing compilation issues because the property was unexpected.